### PR TITLE
remove simple rules

### PR DESCRIPTION
**Pull request by richardstartin**
* Moves setting the error flag to when errors are encountered in decorators, but intercepts the error tag being set
* Moves the HTTP status error rules to the `HttpServerDecorator` and `HttpClientDecorator`s
* Makes the database resource name/statement handling explicit in the decorators, moves the mongodb statement special case in to the `MongoClientDecorator`

### CODE REVIEW

**Review by bantonsson — APPROVED** [review]

File: dd-trace-core/src/test/groovy/datadog/trace/core/processor/TraceProcessorTest.groovy [hunk 1]
```diff
@@ -53,41 +52,6 @@ class TraceProcessorTest extends DDSpecification {
     span.getResourceName() == "404"
   }
 
-  def "#type status #status code error=#error"() {
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/core/processor/TraceProcessorTest.groovy:56]: This test was moved to the decorators

File: dd-trace-core/src/test/groovy/datadog/trace/core/taginterceptor/TagInterceptorTest.groovy [hunk 2]
```diff
@@ -309,23 +309,6 @@ class TagInterceptorTest extends DDSpecification {
     DDTags.MANUAL_DROP | "asdf"  | null
   }
 
-  def "DBStatementAsResource should not interact on Mongo queries"() {
-    when:
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/core/taginterceptor/TagInterceptorTest.groovy:313]: This is tested in the mongo instrumentation

**Comment by richardstartin** [comment]
Something to note here is that there used to be a window of time between setting a tag and reporting where a user could grab the tags map and observe the transient state, but no tests failed as a result of changing this behaviour. The only modified tests were decorator interactions (`setTag` -> specific method), pure input/output of rules which no longer exist, and the removed test logic is either verified by the instrumentation (Mongo) or in decorator tests (HTTP status error mapping). I would argue the transient observable state of the tags map was never specified.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java [hunk 3]
```diff
@@ -354,13 +356,31 @@ public void setMetric(final CharSequence key, final Number value) {
    * @param tag the tag-name
    * @param value the value of the tag. tags with null values are ignored.
    */
-  public void setTag(final String tag, final Object value) {
+  public void setTag(final String tag, final CharSequence value) {
```
Comment (tylerbenson) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:359]: 👍

**Review by tylerbenson — APPROVED** [review]

**Pull request closed without merging**